{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get HLA imputation table\n",
    "## For use in regression, we need to get observations X (N, H) matrices, and Y (N, T) matrices.\n",
    "####  N is the number of participants in our study. H can be one of three things: 4-digit MHC haplotypes, SNPs in the MHC region, and AAs in the classical MHC genes (e.g. DQB1). T is the number of V/J genes whose expression we are interested in Thus X holds genotypes for every individual, and Y holds the expression values of V/J genes for every individual."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "import statsmodels.api as sm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: setup Y (N,T) response matrix from expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "vdj_df = pd.read_csv(\"DGN_vdj_usages.csv\")\n",
    "vdj_table = pd.read_csv(\"DGN_vdj_usages_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"\"\"ONLY GET TRAJ, TRAV, TRBV\"\"\"\n",
    "# query = \"TRA\"\n",
    "# vdj_df = vdj_df[vdj_df['family'].str.startswith(query)]\n",
    "# TODO: also filter vdj_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## plotting usage across families\n",
    "# f, ax = plt.subplots(figsize=(5,30))\n",
    "# sns.boxplot(y=\"family\", x=\"usage_ratio\", data=vdj_df)\n",
    "# plt.xlim((0,0.5))\n",
    "# plt.title(\"Non Protein Coding OR Non Functional Genes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 2: Setup X (N, H) observation matrix from imputed genotype data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### In the bim file output of the SNP2HLA program, there are the following types of names of entities:\n",
    "\n",
    "rs1611715 - intergenic SNP\n",
    "\n",
    "1kg_6_xxxxxx - intergenic SNP from 1000 Genomes\n",
    "\n",
    "SNP_A_30018316 - biallelic SNP within MHC gene\n",
    "\n",
    "SNP_A_30018461_G - multiallelic SNP within MHC gene\n",
    "\n",
    "AA_A_-22_30018317_V 0 multiallelis AA within MHC gene (can have multiple AAs conjugated, or an 'x' for deletion)\n",
    "\n",
    "HLA_A_29 / HLA_A_2901 - binary present/absent of MHC haplotype\n",
    "\n",
    "in the minor/major allele columns, for binary items we have P/A, and for multiallelic we have either the DNA letters or the AA letters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "bim_path = \"/labs/mignot/DGN/Genotypes_DGN/DGNGeno.bim\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_path = \"/labs/mignot/DGN/Genotypes_DGN/DGNGeno.fam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "dosage_path = \"/labs/mignot/DGN/Genotypes_DGN/DGNGeno.dosage\" # each person number between 0 and 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "fam_df = pd.read_csv(fam_path, header=None, delim_whitespace=True) # the column 5 probably is depression/not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_fam_df = fam_df[fam_df[0].apply(lambda name: name.split(\"_\")[-1].startswith(\"LD\"))] # the first 937 patients start with \"LD\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "INCLUDE_INDICES = np.arange(0,937) # indices of people to include in our study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(dosage_path, header=None, delimiter=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_names = list(fam_df[0].apply(lambda name: name.split(\"_\")[-1]).values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [\"genotype\", \"minor\", \"major\"] + subject_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.iloc[:,:3+937] # filter out the non LD people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = [\"genotype\", \"minor\", \"major\"] + [name for name in list(df.columns[3:])] # convert LDxxxx to xxxx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>minor</th>\n",
       "      <th>major</th>\n",
       "      <th>LD0014</th>\n",
       "      <th>LD0041</th>\n",
       "      <th>LD0038</th>\n",
       "      <th>LD0084</th>\n",
       "      <th>LD0022</th>\n",
       "      <th>LD0033</th>\n",
       "      <th>LD0008</th>\n",
       "      <th>...</th>\n",
       "      <th>LD1144</th>\n",
       "      <th>LD1148</th>\n",
       "      <th>LD1115</th>\n",
       "      <th>LD1357</th>\n",
       "      <th>LD1291</th>\n",
       "      <th>LD1282</th>\n",
       "      <th>LD1271</th>\n",
       "      <th>LD1252</th>\n",
       "      <th>LD0165</th>\n",
       "      <th>LD0102</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>rs969931</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.658</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.962</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.961</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.961</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>rs2745406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.658</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.962</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.960</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.997</td>\n",
       "      <td>1.960</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>rs6939431</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.287</td>\n",
       "      <td>0.035</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.287</td>\n",
       "      <td>0.081</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.081</td>\n",
       "      <td>1.008</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>rs1233427</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.643</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.952</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.951</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.951</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>rs1233426</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.643</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.950</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.951</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.951</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 940 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    genotype minor major  LD0014  LD0041  LD0038  LD0084  LD0022  LD0033  \\\n",
       "0   rs969931     C     A   1.993   1.993   1.994   1.993   1.658   1.994   \n",
       "1  rs2745406     C     T   1.993   1.994   1.994   1.994   1.658   1.994   \n",
       "2  rs6939431     A     G   0.015   0.081   0.287   0.035   0.021   0.287   \n",
       "3  rs1233427     A     G   1.993   1.993   1.994   1.993   1.643   1.994   \n",
       "4  rs1233426     A     G   1.993   1.993   1.994   1.993   1.643   1.994   \n",
       "\n",
       "   LD0008  ...  LD1144  LD1148  LD1115  LD1357  LD1291  LD1282  LD1271  \\\n",
       "0   1.993  ...   1.993   1.993   1.962   1.993   1.961   1.993   1.996   \n",
       "1   1.994  ...   1.993   1.994   1.962   1.994   1.960   1.994   1.997   \n",
       "2   0.081  ...   0.015   0.081   0.009   0.081   0.073   0.081   1.008   \n",
       "3   1.993  ...   1.993   1.993   1.952   1.993   1.951   1.993   1.996   \n",
       "4   1.993  ...   1.993   1.993   1.950   1.993   1.951   1.993   1.996   \n",
       "\n",
       "   LD1252  LD0165  LD0102  \n",
       "0   1.961   1.993   1.993  \n",
       "1   1.960   1.994   1.994  \n",
       "2   0.073   0.081   0.081  \n",
       "3   1.951   1.993   1.993  \n",
       "4   1.951   1.993   1.993  \n",
       "\n",
       "[5 rows x 940 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>minor</th>\n",
       "      <th>major</th>\n",
       "      <th>LD0001</th>\n",
       "      <th>LD0002</th>\n",
       "      <th>LD0003</th>\n",
       "      <th>LD0006</th>\n",
       "      <th>LD0007</th>\n",
       "      <th>LD0008</th>\n",
       "      <th>LD0009</th>\n",
       "      <th>...</th>\n",
       "      <th>LD1349</th>\n",
       "      <th>LD1350</th>\n",
       "      <th>LD1353</th>\n",
       "      <th>LD1354</th>\n",
       "      <th>LD1356</th>\n",
       "      <th>LD1357</th>\n",
       "      <th>LD1361</th>\n",
       "      <th>LD1362</th>\n",
       "      <th>LD1364</th>\n",
       "      <th>LD1366</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>rs969931</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.960</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.962</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>rs2745406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.997</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.994</td>\n",
       "      <td>...</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.961</td>\n",
       "      <td>1.997</td>\n",
       "      <td>1.962</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>rs6939431</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.219</td>\n",
       "      <td>1.008</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.287</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.081</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.073</td>\n",
       "      <td>1.008</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.015</td>\n",
       "      <td>0.081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>rs1233427</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.950</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.950</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>rs1233426</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.994</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>...</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.950</td>\n",
       "      <td>1.996</td>\n",
       "      <td>1.950</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "      <td>1.993</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 898 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    genotype minor major  LD0001  LD0002  LD0003  LD0006  LD0007  LD0008  \\\n",
       "0   rs969931     C     A   1.993   1.994   1.996   1.993   1.994   1.993   \n",
       "1  rs2745406     C     T   1.994   1.994   1.997   1.993   1.994   1.994   \n",
       "2  rs6939431     A     G   0.081   0.219   1.008   0.015   0.287   0.081   \n",
       "3  rs1233427     A     G   1.993   1.993   1.996   1.993   1.994   1.993   \n",
       "4  rs1233426     A     G   1.993   1.993   1.996   1.993   1.994   1.993   \n",
       "\n",
       "   LD0009  ...  LD1349  LD1350  LD1353  LD1354  LD1356  LD1357  LD1361  \\\n",
       "0   1.993  ...   1.993   1.993   1.960   1.996   1.962   1.993   1.993   \n",
       "1   1.994  ...   1.994   1.993   1.961   1.997   1.962   1.994   1.993   \n",
       "2   0.081  ...   0.081   0.015   0.073   1.008   0.009   0.081   0.015   \n",
       "3   1.993  ...   1.993   1.993   1.950   1.996   1.950   1.993   1.993   \n",
       "4   1.993  ...   1.993   1.993   1.950   1.996   1.950   1.993   1.993   \n",
       "\n",
       "   LD1362  LD1364  LD1366  \n",
       "0   1.993   1.993   1.993  \n",
       "1   1.993   1.993   1.994  \n",
       "2   0.015   0.015   0.081  \n",
       "3   1.993   1.993   1.993  \n",
       "4   1.993   1.993   1.993  \n",
       "\n",
       "[5 rows x 898 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pick out the columns that are actually present in our Y matrix\n",
    "wanted_columns = [\"genotype\", \"minor\", \"major\"] + list(vdj_table['patid'])\n",
    "\n",
    "df = df.loc[:, wanted_columns]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_HLA(name):\n",
    "    name_tokens = name.split(\"_\")\n",
    "    if name_tokens[0] == \"HLA\" and len(name_tokens[-1]) == 4: # only want 4-digit HLA codes\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_df = df[df['genotype'].apply(lambda name: is_HLA(name))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>minor</th>\n",
       "      <th>major</th>\n",
       "      <th>LD0001</th>\n",
       "      <th>LD0002</th>\n",
       "      <th>LD0003</th>\n",
       "      <th>LD0006</th>\n",
       "      <th>LD0007</th>\n",
       "      <th>LD0008</th>\n",
       "      <th>LD0009</th>\n",
       "      <th>...</th>\n",
       "      <th>LD1349</th>\n",
       "      <th>LD1350</th>\n",
       "      <th>LD1353</th>\n",
       "      <th>LD1354</th>\n",
       "      <th>LD1356</th>\n",
       "      <th>LD1357</th>\n",
       "      <th>LD1361</th>\n",
       "      <th>LD1362</th>\n",
       "      <th>LD1364</th>\n",
       "      <th>LD1366</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>842</td>\n",
       "      <td>HLA_A_0101</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.001</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>843</td>\n",
       "      <td>HLA_A_0102</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>844</td>\n",
       "      <td>HLA_A_0103</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>0.020</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.022</td>\n",
       "      <td>0.045</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014</td>\n",
       "      <td>0.012</td>\n",
       "      <td>0.323</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.170</td>\n",
       "      <td>0.180</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>846</td>\n",
       "      <td>HLA_A_0201</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>1.886</td>\n",
       "      <td>1.892</td>\n",
       "      <td>1.728</td>\n",
       "      <td>0.970</td>\n",
       "      <td>1.874</td>\n",
       "      <td>1.874</td>\n",
       "      <td>1.756</td>\n",
       "      <td>...</td>\n",
       "      <td>0.937</td>\n",
       "      <td>0.934</td>\n",
       "      <td>0.286</td>\n",
       "      <td>0.986</td>\n",
       "      <td>1.093</td>\n",
       "      <td>1.052</td>\n",
       "      <td>1.103</td>\n",
       "      <td>0.441</td>\n",
       "      <td>0.830</td>\n",
       "      <td>1.892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>847</td>\n",
       "      <td>HLA_A_0202</td>\n",
       "      <td>P</td>\n",
       "      <td>A</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.010</td>\n",
       "      <td>0.019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.053</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.031</td>\n",
       "      <td>0.033</td>\n",
       "      <td>0.030</td>\n",
       "      <td>0.054</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.008</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 898 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genotype minor major  LD0001  LD0002  LD0003  LD0006  LD0007  LD0008  \\\n",
       "842  HLA_A_0101     P     A   0.001   0.000   0.002   0.000   0.001   0.001   \n",
       "843  HLA_A_0102     P     A   0.000   0.000   0.000   0.000   0.000   0.000   \n",
       "844  HLA_A_0103     P     A   0.020   0.019   0.048   0.008   0.022   0.022   \n",
       "846  HLA_A_0201     P     A   1.886   1.892   1.728   0.970   1.874   1.874   \n",
       "847  HLA_A_0202     P     A   0.009   0.008   0.021   0.003   0.010   0.010   \n",
       "\n",
       "     LD0009  ...  LD1349  LD1350  LD1353  LD1354  LD1356  LD1357  LD1361  \\\n",
       "842   0.001  ...   0.000   0.000   0.008   0.000   0.004   0.004   0.004   \n",
       "843   0.000  ...   0.000   0.000   0.000   0.000   0.000   0.000   0.000   \n",
       "844   0.045  ...   0.014   0.012   0.323   0.002   0.170   0.180   0.169   \n",
       "846   1.756  ...   0.937   0.934   0.286   0.986   1.093   1.052   1.103   \n",
       "847   0.019  ...   0.006   0.005   0.053   0.001   0.031   0.033   0.030   \n",
       "\n",
       "     LD1362  LD1364  LD1366  \n",
       "842   0.008   0.000   0.000  \n",
       "843   0.000   0.000   0.000  \n",
       "844   0.330   0.021   0.019  \n",
       "846   0.441   0.830   1.892  \n",
       "847   0.054   0.005   0.008  \n",
       "\n",
       "[5 rows x 898 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HLA_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop rows where all HLA haplotype call summed up to 0 for everyone. Nobody had these genotypes.\n",
    "\n",
    "HLA_df = HLA_df.drop(HLA_df[HLA_df.iloc[:, 3:].sum(axis=1) == 0].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "dqb301_arr = np.round(HLA_df[HLA_df['genotype'] == \"HLA_DQB1_0301\"].iloc[:,3:].values[0], 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "dqb602_arr = np.round(HLA_df[HLA_df['genotype'] == \"HLA_DQB1_0602\"].iloc[:,3:].values[0], 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "summed_arr = dqb301_arr + dqb602_arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.22681564245810057"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(summed_arr == 2)/len(summed_arr) # proportion of people with both 0301 and 0602"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_names = HLA_df['genotype'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_df = HLA_df.drop(['minor', 'major'], axis=1).set_index(\"genotype\").transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_df = HLA_df.reset_index().rename(columns={'index':'patid'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert list(HLA_df['patid']) == list(vdj_table['patid'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "HLA_df.to_csv(\"DGN_HLA_df.csv\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
