{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processes the .out files generated by BLAST to get gene counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import pickle\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/labs/mignot/DGN/BLAST\"\n",
    "zero_path = os.path.join(path, \"zero\")\n",
    "one_path = os.path.join(path, \"one\")\n",
    "two_path = os.path.join(path, \"two\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# f = open(\"/home/ashteng/allJCodes.txt\", \"r\")\n",
    "# all_j_codes = set([x.strip() for x in f.readlines()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_name_path_dict = {\"GG\": zero_path} #\"GG\": zero_path, \"GC\": one_path, \n",
    "genotype_name_dose_dict = {\"GG\": 0, \"GC\": 1, \"CC\": 2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GG\n",
      "LD1157.out\n"
     ]
    }
   ],
   "source": [
    "#df = pd.DataFrame(columns=[\"patid\", \"genotype\", \"dose\", \"transcript\", \"count\"])\n",
    "out_file = open(\"VDJCounts_GG.txt\", \"w\")\n",
    "out_file.write(\"patid,genotype,dose,transcript,count\\n\")\n",
    "for genotype in genotype_name_path_dict.keys(): # zero, one, two\n",
    "    print(genotype)\n",
    "    genotype_path = genotype_name_path_dict[genotype]\n",
    "    for file in os.listdir(genotype_path): # for each person\n",
    "        print(file)\n",
    "        patid, extension = file[:-4], file[-4:]\n",
    "        if extension == \".out\":\n",
    "            if os.path.exists(\"{}.fasta\".format(patid)): # skip .out files where the fasta is still present, unfinished\n",
    "                continue \n",
    "            filepath = os.path.join(genotype_path, file)\n",
    "            # load each person's .out file into a separate dataframe\n",
    "            raw_df = pd.read_csv(filepath, delimiter=\"\\t\", names=['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore'])\n",
    "            # group by read id and only extract the maximum score read across families\n",
    "            # we only want one hit for each RNAseq transcript (61584 rows --> 18868 rows)\n",
    "            raw_df = raw_df.loc[raw_df.groupby([\"qseqid\"], sort=False)[\"pident\"].idxmax()]\n",
    "            # for each row in an .out file, we only care about which transcript was hit (i.e. sseqid).\n",
    "            # so for each person, we count all of the sseqid hits from unique transcripts and put into a dictionary\n",
    "            transcript_count_dict = raw_df.groupby(\"sseqid\")['sseqid'].count().to_dict()\n",
    "            for transcript in transcript_count_dict:\n",
    "                out_file.write(\"{},{},{},{},{}\\n\".format(patid,genotype,genotype_name_dose_dict[genotype],transcript,transcript_count_dict[transcript]))\n",
    "                #df = df.append({\"patid\": patid, \"genotype\": genotype, \"dose\": genotype_name_dose_dict[genotype], \"transcript\": transcript, \"count\": transcript_count_dict[transcript]}, ignore_index=True)\n",
    "        break\n",
    "# #             all_j_codes_dict = {x:0 for x in all_j_codes} # each person has a counts distribution over all J codes\n",
    "# #             for index, row in raw_df.iterrows():\n",
    "# #                 match = row['sseqid']\n",
    "# #                 if match in all_j_codes_dict:\n",
    "# #                     all_j_codes_dict[match] += 1\n",
    "# #                 else:\n",
    "# #                     print(\"Error: unknown match {} from {}.\".format(match, filepath))\n",
    "# #             for j_code in all_j_codes_dict:\n",
    "# #                 df = df.append({\"patid\": patid, \"genotype\": genotype, \"dose\": genotype_name_dose_dict[genotype], \"transcript\": j_code, \"count\": all_j_codes_dict[j_code]}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(all_j_codes_dict) == len(df['transcript'].unique()) == len(df[df['patid'] == 'LD0420'])\n",
    "df.to_csv(\"allJCounts2.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"allJCounts2.txt\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['transcript'].str.startswith(\"TRA\")] # only keep TRAJ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_family_subtype_df = df[\"transcript\"].str.split(\"*\", n = 1, expand = True).rename(columns={0:\"family\", 1:\"subtype\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"family\"] = transcript_family_subtype_df[\"family\"]\n",
    "df[\"subtype\"] = transcript_family_subtype_df[\"subtype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For comparison with Sharon\n",
    "patid_family_counts_df = df[['patid', 'family', 'count']].groupby(['patid', 'family']).sum().reset_index()\n",
    "pickle.dump(patid_family_counts_df, open(\"patid_family_count_df.pkl\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patid_genotype_df = df[['patid', 'genotype', 'dose']].drop_duplicates()\n",
    "patid_genotype_df.head()\n",
    "patid_genotype_df.to_csv(\"patid_genotype.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.patid.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_j_count_df = df[[\"patid\", \"count\"]].groupby(\"patid\").sum().rename(columns={\"count\":\"total_J_count\"}).reset_index()\n",
    "total_j_count_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total = pd.merge(df, total_j_count_df, how=\"inner\", on=\"patid\")\n",
    "df_with_total[\"ratio\"] = df_with_total[\"count\"]/df_with_total[\"total_J_count\"]\n",
    "df_with_total['patid'] = df_with_total['patid'].apply(lambda x: int(x[2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_j_ratios_df = df_with_total[[\"patid\", \"transcript\", \"ratio\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_j_ratios_df.to_csv(\"allJRatios.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add covariates to DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCA_path = \"/labs/mignot/GenRED.II.autosomalClean.Final.mds\"\n",
    "PCA_df = pd.read_csv(PCA_path, delim_whitespace=True)\n",
    "PCA_df = PCA_df.drop(\"IID\", axis=1)\n",
    "PCA_df['patid'] = PCA_df['FID'].apply(lambda x: x.split(\"_\")[-1])\n",
    "PCA_df = PCA_df[PCA_df['patid'].str.startswith(\"LD\")]\n",
    "PCA_df['patid'] = PCA_df['patid'].apply(lambda x: int(x[2:]))\n",
    "PCA_df = PCA_df.drop([\"FID\", \"SOL\"], axis=1)\n",
    "PCA_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patid_genotype_df['patid'] = patid_genotype_df['patid'].apply(lambda x: int(x[2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPlotDf(query_transcripts, plot_column=\"ratio\"):\n",
    "    # query_transcripts can be single, or list of multiple, which will be summed together\n",
    "    summed_plot_column_name = \"summed_{}\".format(plot_column)\n",
    "    \n",
    "    # filter out rows that match query_transcripts\n",
    "    filtered_transcripts_df = df_with_total[df_with_total['transcript'].isin(query_transcripts)]\n",
    "    \n",
    "    # sum up all row values by groupby patid\n",
    "    summed_filtered_transcripts_df = filtered_transcripts_df[['patid', plot_column]].groupby('patid').sum().reset_index().rename(columns={plot_column:summed_plot_column_name})\n",
    "    \n",
    "    # add metadata about genotypes and dose\n",
    "    plot_df = pd.merge(summed_filtered_transcripts_df, patid_genotype_df)\n",
    "    \n",
    "    # normalize summed_count column within each class, creating new column normalized_summed_count\n",
    "    normalized_summed_plot_column_name = 'normalized_{}'.format(summed_plot_column_name)\n",
    "    plot_df[normalized_summed_plot_column_name] = (plot_df[summed_plot_column_name] - plot_df[summed_plot_column_name].mean()) / plot_df[summed_plot_column_name].std()\n",
    "\n",
    "    # remove outliers (two sides) using the normalized_summed_count\n",
    "    quantile_low, quantile_high = plot_df[normalized_summed_plot_column_name].quantile([0.025, 0.975]).values\n",
    "    outliers_removed_plot_df = plot_df[(plot_df[normalized_summed_plot_column_name] > quantile_low) & (plot_df[normalized_summed_plot_column_name] < quantile_high)]\n",
    "    \n",
    "#     # remove outliers with hard threshold\n",
    "#     outliers_removed_plot_df = plot_df[plot_df[summed_plot_column_name] <= 0.025]\n",
    "    \n",
    "    # join in covariates\n",
    "    plot_df_with_covariates = pd.merge(outliers_removed_plot_df, PCA_df, on=\"patid\", how=\"inner\")\n",
    "    \n",
    "    return plot_df_with_covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drawFamilyBoxPlots(plot_df, plot_column_name):    \n",
    "    fig, ax = plt.subplots()\n",
    "    sns.boxplot(x=\"genotype\", y=plot_column_name, data=plot_df, order=[\"GG\", \"GC\", \"CC\"], ax=ax)\n",
    "    sns.regplot(x=\"dose\", y=plot_column_name, data=plot_df, x_estimator=np.median, color=\"yellow\", ax=ax, label=\"Median Estimator\")\n",
    "    #sns.scatterplot(x=\"dose\", y=summed_plot_column, data=plot_df)\n",
    "    plt.legend()\n",
    "    plt.title(\"BLAST {} {}\".format(query_transcripts, plot_column_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drawFamilyViolinPlots(plot_df, plot_column_name):\n",
    "    fig, ax = plt.subplots()\n",
    "    sns.violinplot(x=\"genotype\", y=plot_column_name, data=plot_df)\n",
    "    plt.title(\"BLAST {} {}\".format(query_transcripts, plot_column_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def regress(plot_df, plot_column_name, covariates=True):\n",
    "    if covariates:\n",
    "        X = plot_df.loc[:, [\"dose\", \"C1\", \"C2\", \"C3\", \"C4\"]].values\n",
    "    else:\n",
    "        X = plot_df.loc[:, [\"dose\"]].values\n",
    "    y = plot_df[plot_column_name].values\n",
    "    XX = sm.add_constant(X)\n",
    "    mod = sm.OLS(y, XX)\n",
    "    res = mod.fit()\n",
    "    print('Parameters: ', res.params)\n",
    "    print('R2: ', res.rsquared)\n",
    "    print('p values:', res.pvalues)\n",
    "    print(res.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.drop_duplicates('patid').groupby(\"genotype\").count() # final cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_transcripts = [\"TRAJ24*01\"]\n",
    "plot_df = getPlotDf(query_transcripts, \"ratio\")\n",
    "plot_column_name = \"summed_ratio\"\n",
    "regress(plot_df, plot_column_name)\n",
    "drawFamilyBoxPlots(plot_df, plot_column_name)\n",
    "drawFamilyViolinPlots(plot_df, plot_column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_transcripts = [\"TRAJ28*01\"]\n",
    "plot_df = getPlotDf(query_transcripts, \"ratio\")\n",
    "plot_column_name = \"summed_ratio\"\n",
    "regress(plot_df, plot_column_name)\n",
    "drawFamilyBoxPlots(plot_df, plot_column_name)\n",
    "drawFamilyViolinPlots(plot_df, plot_column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non Linear Model Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GG_arr = plot_df[plot_df['genotype'] == \"GG\"]['summed_ratio']\n",
    "GC_arr = plot_df[plot_df['genotype'] == \"GC\"]['summed_ratio']\n",
    "CC_arr = plot_df[plot_df['genotype'] == \"CC\"]['summed_ratio']\n",
    "haveC_arr = np.concatenate((GC_arr.values, CC_arr.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, CC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, GC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GC_arr, CC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, haveC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
