{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import pickle\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"/labs/mignot/DGN/BLAST\"\n",
    "zero_path = os.path.join(path, \"zero\")\n",
    "one_path = os.path.join(path, \"one\")\n",
    "two_path = os.path.join(path, \"two\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# f = open(\"/home/ashteng/allJCodes.txt\", \"r\")\n",
    "# all_j_codes = set([x.strip() for x in f.readlines()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_name_path_dict = {\"GG\": zero_path} #\"GG\": zero_path, \"GC\": one_path, \n",
    "genotype_name_dose_dict = {\"GG\": 0, \"GC\": 1, \"CC\": 2}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GG\n",
      "LD1157.out\n",
      "LD0105.out\n",
      "LD1145.out\n",
      "LD0806.out\n",
      "LD0811.out\n",
      "LD0993.out\n",
      "LD0561.out\n",
      "LD1044.out\n",
      "LD0070.out\n",
      "LD0270.out\n",
      "LD1283.out\n",
      "LD0313.out\n",
      "LD1110.out\n",
      "LD0442.out\n",
      "LD0427.out\n",
      "LD1092.out\n",
      "LD0027.out\n",
      "LD1127.out\n",
      "LD0730.out\n",
      "LD0148.out\n",
      "LD0579.out\n",
      "LD0826.out\n",
      "LD1306.out\n",
      "LD1020.out\n",
      "LD1309.out\n",
      "LD0275.out\n",
      "LD0512.out\n",
      "LD0122.out\n",
      "LD0347.out\n",
      "LD1305.out\n",
      "LD0134.out\n",
      "LD0061.out\n",
      "LD0205.out\n",
      "LD0833.out\n",
      "LD1285.out\n",
      "LD0531.out\n",
      "LD1193.out\n",
      "LD0368.out\n",
      "LD0041.out\n",
      "LD0171.out\n",
      "LD0934.out\n",
      "LD0009.out\n",
      "LD1131.out\n",
      "LD1225.out\n",
      "LD0044.out\n",
      "LD0873.out\n",
      "LD0413.out\n",
      "LD1159.out\n",
      "LD0158.out\n",
      "LD1214.out\n",
      "LD0468.out\n",
      "LD0388.out\n",
      "LD0189.out\n",
      "LD1248.out\n",
      "LD0706.out\n",
      "LD1242.out\n",
      "LD0689.out\n",
      "LD0140.out\n",
      "LD0659.out\n",
      "LD0745.out\n",
      "LD0219.out\n",
      "LD1134.out\n",
      "LD1100.out\n",
      "LD0404.out\n",
      "LD0135.out\n",
      "LD0879.out\n",
      "LD0430.out\n",
      "LD0146.out\n",
      "LD1142.out\n",
      "LD0573.out\n",
      "LD1176.out\n",
      "LD0456.out\n",
      "LD0741.out\n",
      "LD0467.out\n",
      "LD0420.out\n",
      "LD0132.out\n",
      "LD0452.out\n",
      "LD0633.out\n",
      "LD1249.out\n",
      "LD0304.out\n",
      "LD0625.out\n",
      "LD0236.out\n",
      "LD1154.out\n",
      "LD1200.out\n",
      "LD1006.out\n",
      "LD0303.out\n",
      "LD1252.out\n",
      "LD1332.out\n",
      "LD1286.out\n",
      "LD0970.out\n",
      "LD0515.out\n",
      "LD0941.out\n",
      "LD0365.out\n",
      "LD0571.out\n",
      "LD0133.out\n",
      "LD0265.out\n",
      "LD0314.out\n",
      "LD0141.out\n",
      "LD0917.out\n",
      "LD1189.out\n",
      "LD0955.out\n",
      "LD0494.out\n",
      "LD0725.out\n",
      "LD0160.out\n",
      "LD0480.out\n",
      "LD0929.out\n",
      "LD0195.out\n",
      "LD1161.out\n",
      "LD0282.out\n",
      "LD1083.out\n",
      "LD1115.out\n",
      "LD0798.out\n",
      "LD0548.out\n",
      "LD0994.out\n",
      "LD0774.out\n",
      "LD0459.out\n",
      "LD0230.out\n",
      "LD0302.out\n",
      "LD0440.out\n",
      "LD0434.out\n",
      "LD0306.out\n",
      "LD0749.out\n",
      "LD1046.out\n",
      "LD0946.out\n",
      "LD0793.out\n",
      "LD0284.out\n",
      "LD0326.out\n",
      "LD1170.out\n",
      "LD1275.out\n",
      "LD0294.out\n",
      "LD0540.out\n",
      "LD0144.out\n",
      "LD1143.out\n",
      "LD0378.out\n",
      "LD1202.out\n",
      "LD0724.out\n",
      "LD0563.out\n",
      "LD0423.out\n",
      "LD0791.out\n",
      "LD0231.out\n",
      "LD0997.out\n",
      "LD0953.out\n",
      "LD0913.out\n",
      "LD0626.out\n",
      "LD0634.out\n",
      "LD0584.out\n",
      "LD1028.out\n",
      "LD0063.out\n",
      "LD1175.out\n",
      "LD0251.out\n",
      "LD0572.out\n",
      "LD1322.out\n",
      "LD1171.out\n",
      "LD1274.out\n",
      "LD0418.out\n",
      "LD0341.out\n",
      "LD1073.out\n",
      "LD1239.out\n",
      "LD0216.out\n",
      "LD1212.out\n",
      "LD0631.out\n",
      "LD0371.out\n",
      "LD0797.out\n",
      "LD0345.out\n",
      "LD0711.out\n",
      "LD0594.out\n",
      "LD0289.out\n",
      "LD0235.out\n",
      "LD1291.out\n",
      "LD1362.out\n",
      "LD1259.out\n",
      "LD1026.out\n",
      "LD0509.out\n",
      "LD0686.out\n",
      "LD0508.out\n",
      "LD0846.out\n",
      "LD0967.out\n",
      "LD0218.out\n",
      "LD0062.out\n",
      "LD0381.out\n",
      "LD0695.out\n",
      "LD1265.out\n",
      "LD0871.out\n",
      "LD1282.out\n",
      "LD0017.out\n",
      "LD0016.out\n",
      "LD0324.out\n",
      "LD1008.out\n",
      "LD1167.out\n",
      "LD1267.out\n",
      "LD0918.out\n",
      "LD0387.out\n",
      "LD1097.out\n",
      "LD1301.out\n",
      "LD0775.out\n",
      "LD0011.out\n",
      "LD1101.out\n",
      "LD0687.out\n",
      "LD0065.out\n",
      "LD0068.out\n",
      "LD0083.out\n",
      "LD0789.out\n",
      "LD0001.out\n",
      "LD0119.out\n",
      "LD0682.out\n",
      "LD0545.out\n",
      "LD0986.out\n",
      "LD0816.out\n",
      "LD0875.out\n",
      "LD0477.out\n",
      "LD0936.out\n",
      "LD1262.out\n",
      "LD0479.out\n",
      "LD0987.out\n",
      "LD0616.out\n",
      "LD0403.out\n",
      "LD0611.out\n",
      "LD0527.out\n",
      "LD0740.out\n",
      "LD0680.out\n",
      "LD1281.out\n",
      "LD0592.out\n",
      "LD0502.out\n",
      "LD1096.out\n",
      "LD0268.out\n",
      "LD1002.out\n",
      "LD1245.out\n",
      "LD0329.out\n",
      "LD1280.out\n",
      "LD0652.out\n",
      "LD0354.out\n",
      "LD0177.out\n",
      "LD0193.out\n",
      "LD0130.out\n",
      "LD0151.out\n",
      "LD0603.out\n",
      "LD0844.out\n",
      "LD1328.out\n",
      "LD1095.out\n",
      "LD0610.out\n",
      "LD0649.out\n",
      "LD0567.out\n",
      "LD0176.out\n",
      "LD1217.out\n",
      "LD0238.out\n",
      "LD1334.out\n",
      "LD0491.out\n",
      "LD1003.out\n",
      "LD1329.out\n",
      "LD0737.out\n",
      "LD0108.out\n",
      "LD1015.out\n",
      "LD1289.out\n",
      "LD0089.out\n",
      "LD0960.out\n",
      "LD1138.out\n",
      "LD0353.out\n",
      "LD0900.out\n",
      "allJ\n",
      "LD0852.out\n",
      "LD1054.out\n",
      "LD0184.out\n",
      "LD0333.out\n",
      "LD0526.out\n",
      "LD0118.out\n",
      "LD0006.out\n",
      "LD0817.out\n",
      "LD1102.out\n",
      "LD0492.out\n",
      "LD0112.out\n",
      "LD1017.out\n",
      "LD0795.out\n",
      "LD0308.out\n",
      "LD1064.out\n",
      "LD0290.out\n",
      "LD0097.out\n",
      "LD0082.out\n",
      "LD0003.out\n",
      "LD1279.out\n",
      "LD0962.out\n",
      "LD1270.out\n",
      "LD0559.out\n",
      "LD0120.out\n",
      "LD0059.out\n",
      "LD0903.out\n",
      "LD1251.out\n",
      "LD0471.out\n",
      "LD1158.out\n",
      "LD1162.out\n",
      "LD0522.out\n",
      "LD0621.out\n",
      "LD0150.out\n",
      "LD0697.out\n",
      "LD1182.out\n",
      "LD1277.out\n",
      "LD0161.out\n",
      "LD0876.out\n",
      "LD0228.out\n",
      "LD1061.out\n",
      "LD0996.out\n",
      "LD1220.out\n",
      "LD0187.out\n",
      "LD1219.out\n",
      "LD1132.out\n",
      "LD0870.out\n",
      "LD0393.out\n",
      "LD0699.out\n",
      "LD0058.out\n",
      "LD0436.out\n",
      "LD0667.out\n",
      "LD0818.out\n",
      "LD1186.out\n",
      "LD1081.out\n",
      "LD0784.out\n",
      "LD0426.out\n",
      "LD0127.out\n",
      "LD0020.out\n",
      "LD0562.out\n",
      "LD0429.out\n",
      "LD0094.out\n",
      "LD0624.out\n",
      "LD0663.out\n",
      "LD0220.out\n",
      "LD0498.out\n",
      "LD0641.out\n",
      "LD1166.out\n",
      "LD0454.out\n",
      "LD0482.out\n",
      "LD0635.out\n",
      "LD0503.out\n",
      "LD1208.out\n",
      "LD0019.out\n",
      "LD0564.out\n",
      "LD1209.out\n",
      "LD0703.out\n",
      "LD0472.out\n",
      "LD0586.out\n",
      "LD0283.out\n",
      "LD1188.out\n",
      "LD0124.out\n",
      "LD1024.out\n",
      "LD0411.out\n",
      "LD0495.out\n",
      "LD1272.out\n",
      "LD1321.out\n",
      "LD0277.out\n",
      "LD0629.out\n",
      "LD0710.out\n",
      "LD0843.out\n",
      "LD0851.out\n",
      "LD1323.out\n",
      "LD0942.out\n",
      "LD0069.out\n",
      "LD0820.out\n",
      "LD1199.out\n",
      "LD1047.out\n",
      "LD1000.out\n",
      "LD0024.out\n",
      "LD1341.out\n",
      "LD1087.out\n",
      "LD0969.out\n",
      "LD0857.out\n",
      "LD0463.out\n",
      "LD0975.out\n",
      "LD0473.out\n",
      "LD0944.out\n",
      "LD1104.out\n",
      "LD0507.out\n",
      "LD0483.out\n",
      "LD1343.out\n",
      "LD1141.out\n",
      "LD0552.out\n",
      "LD0162.out\n",
      "LD0558.out\n",
      "LD0369.out\n",
      "LD1179.out\n",
      "LD0207.out\n",
      "LD0827.out\n",
      "LD0181.out\n",
      "LD0593.out\n",
      "LD0825.out\n",
      "LD0894.out\n",
      "LD1185.out\n",
      "LD0348.out\n",
      "LD1180.out\n",
      "LD0088.out\n",
      "LD0307.out\n",
      "LD0038.out\n",
      "LD0364.out\n",
      "LD0511.out\n",
      "LD0432.out\n",
      "LD1314.out\n",
      "LD0609.out\n",
      "LD0421.out\n",
      "LD1228.out\n",
      "LD0441.out\n",
      "LD1125.out\n",
      "LD0947.out\n",
      "LD0582.out\n",
      "LD0243.out\n",
      "LD0965.out\n",
      "LD0765.out\n",
      "LD0042.out\n",
      "LD1034.out\n",
      "LD0884.out\n",
      "LD0980.out\n",
      "LD1077.out\n",
      "LD0056.out\n",
      "LD0760.out\n",
      "LD0233.out\n",
      "LD0154.out\n",
      "LD0380.out\n",
      "LD0677.out\n",
      "LD1150.out\n",
      "LD0805.out\n",
      "LD0045.out\n",
      "LD0523.out\n",
      "LD1031.out\n",
      "LD1235.out\n",
      "LD1331.out\n",
      "LD0654.out\n",
      "LD0109.out\n",
      "LD1022.out\n",
      "LD0612.out\n",
      "LD0269.out\n",
      "LD0758.out\n",
      "LD0213.out\n",
      "LD1085.out\n",
      "LD0665.out\n",
      "LD0435.out\n",
      "LD0570.out\n",
      "LD0079.out\n",
      "LD0880.out\n",
      "LD0645.out\n",
      "LD0465.out\n",
      "LD0168.out\n",
      "LD0915.out\n",
      "LD0896.out\n",
      "LD0261.out\n",
      "LD0583.out\n",
      "LD0221.out\n",
      "LD0226.out\n",
      "LD1271.out\n",
      "LD0040.out\n",
      "LD0247.out\n",
      "LD0801.out\n",
      "LD0131.out\n",
      "LD0676.out\n",
      "LD0520.out\n",
      "LD0858.out\n",
      "LD1075.out\n",
      "LD1226.out\n",
      "LD0519.out\n",
      "LD0715.out\n",
      "LD0173.out\n",
      "LD1113.out\n",
      "LD0575.out\n",
      "LD0102.out\n",
      "LD0554.out\n",
      "LD0299.out\n",
      "LD0782.out\n",
      "LD1093.out\n",
      "LD0685.out\n",
      "LD0618.out\n",
      "LD0128.out\n",
      "LD0916.out\n",
      "LD0954.out\n",
      "LD0337.out\n",
      "LD0814.out\n",
      "LD0035.out\n",
      "LD1255.out\n",
      "LD1316.out\n",
      "LD0084.out\n",
      "LD0208.out\n",
      "LD1349.out\n",
      "LD0744.out\n",
      "LD0310.out\n",
      "LD1258.out\n",
      "LD0796.out\n",
      "LD1042.out\n",
      "LD1045.out\n",
      "LD0252.out\n",
      "LD0145.out\n",
      "LD0940.out\n",
      "LD0597.out\n",
      "LD0192.out\n",
      "LD0488.out\n",
      "LD1198.out\n",
      "LD1165.out\n",
      "LD0487.out\n",
      "LD0007.out\n",
      "LD0766.out\n",
      "LD1029.out\n",
      "LD0278.out\n",
      "LD0496.out\n",
      "LD0444.out\n",
      "LD0702.out\n",
      "LD0033.out\n",
      "LD0457.out\n",
      "LD1264.out\n",
      "LD0272.out\n",
      "LD0249.out\n",
      "LD0899.out\n",
      "LD1325.out\n",
      "LD1072.out\n",
      "LD0864.out\n",
      "LD0754.out\n",
      "LD1076.out\n",
      "LD1056.out\n",
      "LD0098.out\n",
      "LD0568.out\n",
      "LD0136.out\n",
      "LD0086.out\n",
      "LD0116.out\n",
      "LD0422.out\n",
      "LD0416.out\n",
      "LD1263.out\n",
      "LD0693.out\n",
      "LD0727.out\n",
      "LD0439.out\n",
      "LD1313.out\n",
      "LD0601.out\n",
      "LD0224.out\n",
      "LD0855.out\n",
      "LD0253.out\n",
      "LD0263.out\n",
      "LD1197.out\n",
      "LD0604.out\n",
      "LD0248.out\n",
      "LD0300.out\n",
      "LD0183.out\n",
      "LD1021.out\n",
      "LD1080.out\n",
      "LD1074.out\n",
      "LD0669.out\n",
      "LD0769.out\n",
      "LD0417.out\n",
      "LD1039.out\n",
      "LD0738.out\n",
      "LD1163.out\n",
      "LD0709.out\n",
      "LD1012.out\n",
      "LD0638.out\n",
      "LD1013.out\n",
      "LD0623.out\n",
      "LD0323.out\n",
      "LD0890.out\n",
      "LD0598.out\n",
      "LD0533.out\n",
      "LD1155.out\n",
      "LD0244.out\n",
      "LD0437.out\n",
      "LD1293.out\n",
      "LD1135.out\n",
      "LD0090.out\n",
      "LD0565.out\n",
      "LD0190.out\n",
      "LD1153.out\n",
      "LD0450.out\n",
      "LD0581.out\n",
      "LD1103.out\n",
      "LD0410.out\n",
      "LD0823.out\n",
      "LD0217.out\n",
      "LD0736.out\n",
      "LD0768.out\n",
      "LD1210.out\n",
      "LD0451.out\n",
      "LD1203.out\n",
      "LD1174.out\n",
      "LD0662.out\n",
      "LD0534.out\n",
      "LD0157.out\n",
      "LD0174.out\n",
      "LD0959.out\n",
      "LD1190.out\n",
      "LD0865.out\n",
      "LD0544.out\n",
      "LD0104.out\n",
      "LD0923.out\n",
      "LD0891.out\n",
      "LD1181.out\n",
      "LD1268.out\n",
      "LD1149.out\n",
      "LD0977.out\n",
      "LD1243.out\n",
      "LD1090.out\n",
      "LD1232.out\n",
      "LD1065.out\n",
      "LD0175.out\n",
      "LD1356.out\n",
      "LD0458.out\n",
      "LD1059.out\n",
      "LD0054.out\n",
      "LD0013.out\n",
      "LD0012.out\n",
      "LD0209.out\n",
      "LD0484.out\n",
      "LD0255.out\n",
      "LD0720.out\n",
      "LD0862.out\n",
      "LD1038.out\n",
      "LD0790.out\n",
      "LD1254.out\n",
      "LD1037.out\n",
      "LD0878.out\n",
      "LD0619.out\n",
      "LD0034.out\n",
      "LD0979.out\n",
      "LD1298.out\n",
      "LD0325.out\n",
      "LD0103.out\n",
      "LD1094.out\n",
      "LD0620.out\n",
      "LD0163.out\n",
      "LD0952.out\n",
      "LD0786.out\n",
      "LD0100.out\n",
      "LD0030.out\n",
      "LD1071.out\n",
      "LD0746.out\n",
      "LD0525.out\n",
      "LD0481.out\n",
      "LD0101.out\n",
      "LD1144.out\n",
      "LD0599.out\n",
      "LD0644.out\n",
      "LD0215.out\n",
      "LD0821.out\n",
      "LD0886.out\n",
      "LD0815.out\n",
      "LD0859.out\n",
      "LD1241.out\n",
      "LD1049.out\n",
      "LD1105.out\n",
      "LD0316.out\n",
      "LD1348.out\n",
      "LD0317.out\n",
      "LD0995.out\n",
      "LD1230.out\n",
      "LD1168.out\n",
      "LD0152.out\n",
      "LD0374.out\n",
      "LD1345.out\n",
      "LD0763.out\n",
      "LD0047.out\n",
      "LD0800.out\n",
      "LD0407.out\n",
      "LD0630.out\n",
      "LD0794.out\n",
      "LD0764.out\n",
      "LD0834.out\n",
      "LD1147.out\n",
      "LD0445.out\n",
      "LD0841.out\n",
      "LD0327.out\n",
      "LD0349.out\n",
      "LD0292.out\n",
      "LD0014.out\n",
      "LD0075.out\n",
      "LD1091.out\n",
      "LD0991.out\n",
      "LD0910.out\n"
     ]
    }
   ],
   "source": [
    "#df = pd.DataFrame(columns=[\"patid\", \"genotype\", \"dose\", \"transcript\", \"count\"])\n",
    "out_file = open(\"VDJCounts_GG.txt\", \"w\")\n",
    "out_file.write(\"patid,genotype,dose,transcript,count\\n\")\n",
    "for genotype in genotype_name_path_dict.keys(): # zero, one, two\n",
    "    print(genotype)\n",
    "    genotype_path = genotype_name_path_dict[genotype]\n",
    "    for file in os.listdir(genotype_path): # for each person\n",
    "        print(file)\n",
    "        patid, extension = file[:-4], file[-4:]\n",
    "        if extension == \".out\":\n",
    "            if os.path.exists(\"{}.fasta\".format(patid)): # skip .out files where the fasta is still present, unfinished\n",
    "                continue \n",
    "            filepath = os.path.join(genotype_path, file)\n",
    "            # load each person's .out file into a separate dataframe\n",
    "            raw_df = pd.read_csv(filepath, delimiter=\"\\t\", names=['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'sstart', 'send', 'evalue', 'bitscore'])\n",
    "            # group by read id and only extract the maximum score read across families\n",
    "            # we only want one hit for each RNAseq transcript (61584 rows --> 18868 rows)\n",
    "            raw_df = raw_df.loc[raw_df.groupby([\"qseqid\"], sort=False)[\"pident\"].idxmax()]\n",
    "            # for each row in an .out file, we only care about which transcript was hit (i.e. sseqid).\n",
    "            # so for each person, we count all of the sseqid hits from unique transcripts and put into a dictionary\n",
    "            transcript_count_dict = raw_df.groupby(\"sseqid\")['sseqid'].count().to_dict()\n",
    "            for transcript in transcript_count_dict:\n",
    "                out_file.write(\"{},{},{},{},{}\\n\".format(patid,genotype,genotype_name_dose_dict[genotype],transcript,transcript_count_dict[transcript]))\n",
    "                #df = df.append({\"patid\": patid, \"genotype\": genotype, \"dose\": genotype_name_dose_dict[genotype], \"transcript\": transcript, \"count\": transcript_count_dict[transcript]}, ignore_index=True)\n",
    "\n",
    "# #             all_j_codes_dict = {x:0 for x in all_j_codes} # each person has a counts distribution over all J codes\n",
    "# #             for index, row in raw_df.iterrows():\n",
    "# #                 match = row['sseqid']\n",
    "# #                 if match in all_j_codes_dict:\n",
    "# #                     all_j_codes_dict[match] += 1\n",
    "# #                 else:\n",
    "# #                     print(\"Error: unknown match {} from {}.\".format(match, filepath))\n",
    "# #             for j_code in all_j_codes_dict:\n",
    "# #                 df = df.append({\"patid\": patid, \"genotype\": genotype, \"dose\": genotype_name_dose_dict[genotype], \"transcript\": j_code, \"count\": all_j_codes_dict[j_code]}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(all_j_codes_dict) == len(df['transcript'].unique()) == len(df[df['patid'] == 'LD0420'])\n",
    "df.to_csv(\"allJCounts2.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"allJCounts2.txt\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['transcript'].str.startswith(\"TRA\")] # only keep TRAJ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_family_subtype_df = df[\"transcript\"].str.split(\"*\", n = 1, expand = True).rename(columns={0:\"family\", 1:\"subtype\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"family\"] = transcript_family_subtype_df[\"family\"]\n",
    "df[\"subtype\"] = transcript_family_subtype_df[\"subtype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For comparison with Sharon\n",
    "patid_family_counts_df = df[['patid', 'family', 'count']].groupby(['patid', 'family']).sum().reset_index()\n",
    "pickle.dump(patid_family_counts_df, open(\"patid_family_count_df.pkl\", \"wb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patid_genotype_df = df[['patid', 'genotype', 'dose']].drop_duplicates()\n",
    "patid_genotype_df.head()\n",
    "patid_genotype_df.to_csv(\"patid_genotype.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(df.patid.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_j_count_df = df[[\"patid\", \"count\"]].groupby(\"patid\").sum().rename(columns={\"count\":\"total_J_count\"}).reset_index()\n",
    "total_j_count_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total = pd.merge(df, total_j_count_df, how=\"inner\", on=\"patid\")\n",
    "df_with_total[\"ratio\"] = df_with_total[\"count\"]/df_with_total[\"total_J_count\"]\n",
    "df_with_total['patid'] = df_with_total['patid'].apply(lambda x: int(x[2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_j_ratios_df = df_with_total[[\"patid\", \"transcript\", \"ratio\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_j_ratios_df.to_csv(\"allJRatios.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add covariates to DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PCA_path = \"/labs/mignot/GenRED.II.autosomalClean.Final.mds\"\n",
    "PCA_df = pd.read_csv(PCA_path, delim_whitespace=True)\n",
    "PCA_df = PCA_df.drop(\"IID\", axis=1)\n",
    "PCA_df['patid'] = PCA_df['FID'].apply(lambda x: x.split(\"_\")[-1])\n",
    "PCA_df = PCA_df[PCA_df['patid'].str.startswith(\"LD\")]\n",
    "PCA_df['patid'] = PCA_df['patid'].apply(lambda x: int(x[2:]))\n",
    "PCA_df = PCA_df.drop([\"FID\", \"SOL\"], axis=1)\n",
    "PCA_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "patid_genotype_df['patid'] = patid_genotype_df['patid'].apply(lambda x: int(x[2:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getPlotDf(query_transcripts, plot_column=\"ratio\"):\n",
    "    # query_transcripts can be single, or list of multiple, which will be summed together\n",
    "    summed_plot_column_name = \"summed_{}\".format(plot_column)\n",
    "    \n",
    "    # filter out rows that match query_transcripts\n",
    "    filtered_transcripts_df = df_with_total[df_with_total['transcript'].isin(query_transcripts)]\n",
    "    \n",
    "    # sum up all row values by groupby patid\n",
    "    summed_filtered_transcripts_df = filtered_transcripts_df[['patid', plot_column]].groupby('patid').sum().reset_index().rename(columns={plot_column:summed_plot_column_name})\n",
    "    \n",
    "    # add metadata about genotypes and dose\n",
    "    plot_df = pd.merge(summed_filtered_transcripts_df, patid_genotype_df)\n",
    "    \n",
    "    # normalize summed_count column within each class, creating new column normalized_summed_count\n",
    "    normalized_summed_plot_column_name = 'normalized_{}'.format(summed_plot_column_name)\n",
    "    plot_df[normalized_summed_plot_column_name] = (plot_df[summed_plot_column_name] - plot_df[summed_plot_column_name].mean()) / plot_df[summed_plot_column_name].std()\n",
    "\n",
    "    # remove outliers (two sides) using the normalized_summed_count\n",
    "    quantile_low, quantile_high = plot_df[normalized_summed_plot_column_name].quantile([0.025, 0.975]).values\n",
    "    outliers_removed_plot_df = plot_df[(plot_df[normalized_summed_plot_column_name] > quantile_low) & (plot_df[normalized_summed_plot_column_name] < quantile_high)]\n",
    "    \n",
    "#     # remove outliers with hard threshold\n",
    "#     outliers_removed_plot_df = plot_df[plot_df[summed_plot_column_name] <= 0.025]\n",
    "    \n",
    "    # join in covariates\n",
    "    plot_df_with_covariates = pd.merge(outliers_removed_plot_df, PCA_df, on=\"patid\", how=\"inner\")\n",
    "    \n",
    "    return plot_df_with_covariates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drawFamilyBoxPlots(plot_df, plot_column_name):    \n",
    "    fig, ax = plt.subplots()\n",
    "    sns.boxplot(x=\"genotype\", y=plot_column_name, data=plot_df, order=[\"GG\", \"GC\", \"CC\"], ax=ax)\n",
    "    sns.regplot(x=\"dose\", y=plot_column_name, data=plot_df, x_estimator=np.median, color=\"yellow\", ax=ax, label=\"Median Estimator\")\n",
    "    #sns.scatterplot(x=\"dose\", y=summed_plot_column, data=plot_df)\n",
    "    plt.legend()\n",
    "    plt.title(\"BLAST {} {}\".format(query_transcripts, plot_column_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drawFamilyViolinPlots(plot_df, plot_column_name):\n",
    "    fig, ax = plt.subplots()\n",
    "    sns.violinplot(x=\"genotype\", y=plot_column_name, data=plot_df)\n",
    "    plt.title(\"BLAST {} {}\".format(query_transcripts, plot_column_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def regress(plot_df, plot_column_name, covariates=True):\n",
    "    if covariates:\n",
    "        X = plot_df.loc[:, [\"dose\", \"C1\", \"C2\", \"C3\", \"C4\"]].values\n",
    "    else:\n",
    "        X = plot_df.loc[:, [\"dose\"]].values\n",
    "    y = plot_df[plot_column_name].values\n",
    "    XX = sm.add_constant(X)\n",
    "    mod = sm.OLS(y, XX)\n",
    "    res = mod.fit()\n",
    "    print('Parameters: ', res.params)\n",
    "    print('R2: ', res.rsquared)\n",
    "    print('p values:', res.pvalues)\n",
    "    print(res.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_with_total.drop_duplicates('patid').groupby(\"genotype\").count() # final cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_transcripts = [\"TRAJ24*01\"]\n",
    "plot_df = getPlotDf(query_transcripts, \"ratio\")\n",
    "plot_column_name = \"summed_ratio\"\n",
    "regress(plot_df, plot_column_name)\n",
    "drawFamilyBoxPlots(plot_df, plot_column_name)\n",
    "drawFamilyViolinPlots(plot_df, plot_column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query_transcripts = [\"TRAJ28*01\"]\n",
    "plot_df = getPlotDf(query_transcripts, \"ratio\")\n",
    "plot_column_name = \"summed_ratio\"\n",
    "regress(plot_df, plot_column_name)\n",
    "drawFamilyBoxPlots(plot_df, plot_column_name)\n",
    "drawFamilyViolinPlots(plot_df, plot_column_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Non Linear Model Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GG_arr = plot_df[plot_df['genotype'] == \"GG\"]['summed_ratio']\n",
    "GC_arr = plot_df[plot_df['genotype'] == \"GC\"]['summed_ratio']\n",
    "CC_arr = plot_df[plot_df['genotype'] == \"CC\"]['summed_ratio']\n",
    "haveC_arr = np.concatenate((GC_arr.values, CC_arr.values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, CC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, GC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GC_arr, CC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(GG_arr, haveC_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
